tirocks-sys/src/lib.rs

```diff
@@ -13,20 +13,196 @@ mod bindings {
     }
 }
 
+/// Convert a rocksdb slice to rust slice.
+///
+/// # Safety
+///
+/// Caller should guarantee the data lives as long as the returned slice.
+// Because the layout of rocksdb_Slice is exact the same as &[u8], this function
+// will be optimized away by compiler.
+#[inline]
+pub unsafe fn s<'a>(src: rocksdb_Slice) -> &'a [u8] {
+    std::slice::from_raw_parts(src.data_ as _, src.size_)
+}
+
+impl rocksdb_Status {
+    #[inline]
+    pub fn with_code(code: rocksdb_Status_Code) -> rocksdb_Status {
+        rocksdb_Status {
+            code_: code,
+            subcode_: rocksdb_Status_SubCode::kNone,
+            sev_: rocksdb_Status_Severity::kNoError,
+            state_: std::ptr::null(),
+        }
+    }
+
+    /// Build a status with given errors.
+    ///
+    /// # Panics
+    ///
+    /// `code` should not be `kOK`.
+    #[inline]
+    pub fn with_error(code: rocksdb_Status_Code, state: impl AsRef<[u8]>) -> rocksdb_Status {
+        assert_ne!(code, rocksdb_Status_Code::kOk);
+        let state = state.as_ref();
+        let state_ = if !state.is_empty() {
+            unsafe { crocksdb_to_cplus_array(r(state)) }
+        } else {
+            std::ptr::null()
+        };
+        rocksdb_Status {
+            code_: code,
+            subcode_: rocksdb_Status_SubCode::kNone,
+            sev_: rocksdb_Status_Severity::kNoError,
+            state_,
+        }
+    }
+
+    #[inline]
+    fn clear_state(&mut self) {
+        if self.state_.is_null() {
+            return;
+        }
+        unsafe {
+            crocksdb_free_cplus_array(self.state_);
+            self.state_ = std::ptr::null();
+        }
+    }
+
+    #[inline]
+    pub fn ok(&self) -> bool {
+        self.code_ == rocksdb_Status_Code::kOk
+    }
+
+    #[inline]
+    pub fn state(&self) -> Option<&[u8]> {
+        if self.state_.is_null() {
+            None
+        } else {
+            unsafe { Some(CStr::from_ptr(self.state_).to_bytes()) }
+        }
+    }
+
+    #[inline]
+    pub fn message(&self) -> Result<Option<&str>, Utf8Error> {
+        if self.state_.is_null() {
+            Ok(None)
+        } else {
+            unsafe { CStr::from_ptr(self.state_).to_str().map(Some) }
+        }
+    }
+
+    #[inline]
+    pub fn code(&self) -> rocksdb_Status_Code {
+        self.code_
+    }
+
+    #[inline]
+    pub fn severity(&self) -> rocksdb_Status_Severity {
+        self.sev_
+    }
+
+    #[inline]
+    pub fn set_severity(&mut self, severity: rocksdb_Status_Severity) {
+        self.sev_ = severity;
+    }
+
+    #[inline]
+    pub fn sub_code(&self) -> rocksdb_Status_SubCode {
+        self.subcode_
+    }
+
+    #[inline]
+    pub fn set_sub_code(&mut self, sub_code: rocksdb_Status_SubCode) {
+        self.subcode_ = sub_code;
+    }
+}
+
+impl Drop for rocksdb_Status {
+    #[inline]
+    fn drop(&mut self) {
+        self.clear_state();
+    }
+}
+
+#[macro_export]
+macro_rules! ffi_try {
+    ($func:ident($($arg:expr),+)) => ({
```

Comment:
can't we use `*` instead of `+`?

Reply:
It can't as we need to add a new parameter at the end no matter there is any paremeters. `func(,&mut status)` is wrong.